tools/update_release_json.go

```diff
@@ -0,0 +1,171 @@
+// tool, for which new releases can be documented.
+func knownComponents() []string {
+	return []string{
+		"agent",
+		"agentConfig",
+	}
+}
+
+// component retrievses a pointer to a known component identified by its name,
+// which must match one of the "knownComponents", so that a new release can be
+// appended to it.
+func component(releases *ReleaseJSON, compName string) (*Component, error) {
+	switch compName {
+	case "agent":
+		return &releases.Agent, nil
+	case "agentConfig":
+		return &releases.AgentConfigTool, nil
+	default:
+		return nil, fmt.Errorf("unknown component \"%s\"", compName)
+	}
+}
+
+// isUniqueTag determines whether a particular tag already appears in a
+// component's release history or not.
+func isUniqueTag(comp Component, tag string) bool {
+	for _, release := range comp.ReleaseHistory {
+		if release.Tag == tag {
+			return false
+		}
+	}
+	return true
+}
+
+// usage prints a usage string giving a more detailed explanation about
+// how to use this tool.
+func usage() {
+	fmt.Fprintf(os.Stderr, "This program is used to update a JSON file describing releases of MIG components.\n\n")
+	fmt.Fprintf(os.Stderr, "Note that the -component, -binary and -tag arguments are required.\n\n")
+	flag.PrintDefaults()
+}
+
+func main() {
+	flag.Usage = usage
+
+	releaseJSONPath := flag.String(
+		"releases",
+		path.Join(defaultReleasesDir, releaseJSONFileName),
+		fmt.Sprintf("Path to the %s file to update", releaseJSONFileName))
+	componentName := flag.String(
+		"component",
+		"",
+		fmt.Sprintf(
+			"The name of the component to update. Must be one of: %s",
+			strings.Join(knownComponents(), ", ")))
+	binary := flag.String(
+		"binary",
+		"",
+		"The name of the binary file")
+	tag := flag.String(
+		"tag",
+		"",
+		"A unique tag string identifying the release")
+	notes := flag.String(
+		"notes",
+		"",
+		"Notes describing the release")
+	sha256 := flag.String(
+		"sha256",
+		"",
+		"Hex encoding of the SHA256 hash of the released binary")
+
+	flag.Parse()
+
+	releaseFile, err := os.Open(*releaseJSONPath)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Could not open releases JSON file \"%s\"\n", *releaseJSONPath)
+		os.Exit(1)
+	}
+	defer releaseFile.Close()
+
+	var releaseJSON ReleaseJSON
+	decoder := json.NewDecoder(releaseFile)
+	decodeErr := decoder.Decode(&releaseJSON)
+	if decodeErr != nil {
+		fmt.Fprintf(os.Stderr, "Failed to decode release JSON. Error: %s\n", decodeErr.Error())
+		os.Exit(1)
+	}
+
+	component, err := component(&releaseJSON, *componentName)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Invalid component specified. Error: %s\n", err.Error())
+		os.Exit(1)
+	}
+
+	if *binary == "" {
+		fmt.Fprintf(os.Stderr, "No binary name supplied.\n")
+	}
+
+	if *tag == "" {
+		fmt.Fprintf(os.Stderr, "No tag identifier supplied.\n")
+		os.Exit(1)
+	}
+
+	setLatestRelease(component, *binary, *tag, *notes, *sha256)
+
+	prefix := ""
+	indent := "    "
```

Comment:
Ironically, the 4-space indent used in the JSON is not used anywhere in this source code otherwise, which uses only 2- or 8-. That’s not a criticism, but it does count as a third indenting style in use in a single file, so duly noted.